code/modules/projectiles/guns/syringe_gun.dm

```diff
@@ -95,3 +94,285 @@
+			to_chat(user, "<span class='warning'>[A] is empty!</span>")
+			return TRUE
+		if(length(syringes) + (chambered?.BB ? 1 : 0) == max_syringes)
+			to_chat(user, "<span class='warning'>[src] is full!</span>")
+			return TRUE
+
+		var/total_inserted = 0
+		for(var/obj/item/reagent_containers/syringe/S in container)
+			if((length(syringes) + (chambered?.BB ? 1 : 0)) == max_syringes)
+				// full
+				break
+
+			// only allow empty syringes
+			if(!S.reagents.total_volume)
+				syringes += S
+				S.forceMove(src)
+				total_inserted++
+
+		if(total_inserted)
+			playsound(src, 'sound/weapons/gun_interactions/bulletinsert.ogg', 50, 1)
+			to_chat(user, "<span class='notice'>You load [total_inserted] empty syringes into [src].")
+			// Chamber a syringe.
+			process_chamber()
+
+	else if(istype(A, /obj/item/reagent_containers/syringe))
+		insert_single_syringe(A, user)
+
+	else if(istype(A, /obj/item/reagent_containers))
+		// Loading with chemicals (but not from syringes)
+		var/obj/item/reagent_containers/incoming = A
+		if(!incoming.is_drainable())
+			return
+
+		if(!incoming.reagents.total_volume)
+			to_chat(user, "<span class='warning'>[incoming] is empty!</span>")
+			return
+
+		if(reagents.holder_full())
+			to_chat(user, "<span class='warning'>[src]'s internal reservoir is full!</span>")
+			return
+
+		var/trans = incoming.reagents.trans_to(src, incoming.amount_per_transfer_from_this)
+		to_chat(user, "<span class='notice'>You transfer [trans] unit\s of the solution to [src]'s internal reservoir.</span>")
+		update_loaded_syringe()
+
+		// Reset the reservoir alarm
+		alarmed = FALSE
+
+	else
+		return ..()
+
+// Allow for emptying your deathmix, or for sec to find out what you were dumping into people
+/obj/item/gun/syringe/rapidsyringe/afterattack(atom/target, mob/living/user, flag, params)
+	if(istype(target, /obj/item/reagent_containers) && !istype(target, /obj/item/reagent_containers/syringe))
+		var/obj/item/reagent_containers/destination = target
+
+		if(!destination.is_refillable())
+			return
+
+		if(!reagents.total_volume)
+			to_chat(user, "<span class='warning'>[src]'s internal reservoir is empty!</span>")
+			return
+
+		if(destination.reagents.holder_full())
+			to_chat(user, "<span class='warning'>[destination] is full.</span>")
+			return
+
+		var/transfer_source = "[src]'s internal reservoir"
+		var/transfer_amount
+
+		// Empty directly from the syringe, if it's loaded.
+		// Otherwise, we wouldn't be able to fully empty the gun's chems without firing it.
+		if(chambered?.BB && chambered.BB.reagents.total_volume)
+			transfer_amount = chambered.BB.reagents.trans_to(destination, get_units_per_shot())
+			transfer_source = "the syringe chambered in [src]"
+		else
+			transfer_amount = reagents.trans_to(destination, reservoir_transfer_amount)
+
+		to_chat(user, "<span class='notice'>You drain [transfer_amount] unit\s from [transfer_source] into [destination].</span>")
+		// Refill the syringe
+		update_loaded_syringe()
```

Comment:
Does this mean that if you have syringes in the gun, then trying to unload it would only ever transfer out of the syringe? Since the chambered syringe (and not the internal reservoir) gets emptied if it's present, and the chambered syrigne gets refilled each time after being emptied?

Am i reading this wrong?

Related: would it maybe be easier to not bother with chambered syringes handling at all, and only refill them *right* before the trigger is pulled? (i don't know)

Alternatively: Make a detachable syringe magazine. That should make things more streamlined, at least code-wise (but maybe in-game too).

Reply:
The chambered syringe is kind of an artifact of this tying into gun code. Guns handle what they fire by chambering a bullet and firing that, and I'm mostly following along with the logic defined in the original syringe gun w/r/t that. From what I was able to identify, there isn't really a great hook to tie into for filling syringes (though I think there might be a signal I could implement for that). 

The reason I empty directly from the syringe is that if there's only like 7 units left in the gun while the syringe is set to fire 15u, that 7u will be in the chambered syringe so it all gets sent out when firing. It also gives you a neat way to set how much you want to dispense, so I see that as a bit of a feature as well.

I do like the idea of a magazine, and I was thinking of like syringe speed loaders being a thing. Could look into that a little more when I get the chance.

Reply:
I still think moving things into magazines might be a good idea, but I don't think it'll make it into this PR.